gobblin-modules/gobblin-azkaban/src/main/java/org/apache/gobblin/service/modules/orchestration/AzkabanClient.java

```diff
@@ -402,97 +329,71 @@ public AzkabanExecuteFlowStatus executeFlowWithOptions(
+   * including a list of all the job executions.
    *
    * @param execId execution id to be fetched.
    *
-   * @return The status object which contains success status and all the detailed information of that execution.
+   * @return The status object which contains success status and all the detailed
+   *         information of that execution.
    */
-  public AzkabanFetchExecuteFlowStatus fetchFlowExecution (String execId) {
-    try {
-      refreshSession();
-      List<NameValuePair> nvps = new ArrayList<>();
-      nvps.add(new BasicNameValuePair(AzkabanClientParams.AJAX, "fetchexecflow"));
-      nvps.add(new BasicNameValuePair(AzkabanClientParams.SESSION_ID, this.sessionId));
-      nvps.add(new BasicNameValuePair(AzkabanClientParams.EXECID, execId));
-
-      Header contentType = new BasicHeader(HttpHeaders.CONTENT_TYPE, "application/x-www-form-urlencoded");
-      Header requestType = new BasicHeader("X-Requested-With", "XMLHttpRequest");
-
-      HttpGet httpGet = new HttpGet(url + "/executor?" + URLEncodedUtils.format(nvps, "UTF-8"));
-      httpGet.setHeaders(new Header[]{contentType, requestType});
-
-      CloseableHttpResponse response = this.httpClient.execute(httpGet);
-      try {
-        Map<String, String> map = handleResponse(response);
-        return new AzkabanFetchExecuteFlowStatus(new AzkabanFetchExecuteFlowStatus.Execution(map));
-      } finally {
-        response.close();
-      }
-    } catch (Exception e) {
-      return new AzkabanFetchExecuteFlowStatus("Azkaban client cannot "
-          + "fetch execId " + execId, e);
-    }
-  }
+  public AzkabanFetchExecuteFlowStatus fetchFlowExecution(String execId) throws AzkabanClientException {
+    AzkabanMultiCallables.FetchFlowExecCallable callable =
+        new AzkabanMultiCallables.FetchFlowExecCallable(this, execId);
 
-  private void addFlowParameters(List<NameValuePair> nvps, Map<String, String> flowParams) {
-    if (flowParams != null) {
-      for (Map.Entry<String, String> entry : flowParams.entrySet()) {
-        String key = entry.getKey();
-        String value = entry.getValue();
-        if (StringUtils.isNotBlank(key) && StringUtils.isNotBlank(value)) {
-          log.debug("New flow parameter added:" + key + "-->" + value);
-          nvps.add(new BasicNameValuePair("flowOverride[" + key + "]", value));
-        }
-      }
-    }
+    return runWithRetry(callable, AzkabanFetchExecuteFlowStatus.class);
   }
 
-  private void addFlowOptions(List<NameValuePair> nvps, Map<String, String> flowOptions) {
-    if (flowOptions != null) {
-      for (Map.Entry<String, String> option : flowOptions.entrySet()) {
-        log.debug("New flow option added:" + option .getKey()+ "-->" + option.getValue());
-        nvps.add(new BasicNameValuePair(option.getKey(), option.getValue()));
+  private <T> T runWithRetry(Callable callable, Class<T> cls) throws AzkabanClientException {
+    try {
+      AzkabanClientStatus status = this.retryer.call(callable);
```

Comment:
How will retry help for the situation when Azkaban server reports that a session is expired? The problem is that our refresh interval could be much longer e.g. 12h, but Az server reports that session is invalid after 10 mts. In this case, retry will not be much useful. We will need to catch the exception and forcibly refresh the session.

Reply:
I force a retry now. Please review the latest